include/revng/Pipes/ModelInvalidationRule.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include "revng/Pipeline/Container.h"
+#include "revng/Pipeline/Runner.h"
+#include "revng/Pipeline/Target.h"
+
+namespace revng::pipes {
+
+class ModelInvalidationRule {
+public:
+  ModelInvalidationRule() { getRulesRegistry().push_back(this); }
+
+  static llvm::ArrayRef<ModelInvalidationRule *> getRules() {
+    return getRulesRegistry();
+  }
+
+  virtual void registerInvalidations(const pipeline::ContainerBase &Container,
+                                     pipeline::TargetsList &ToRemove) const = 0;
+
+  static void
+  registerAllInvalidations(const pipeline::Runner &Runner,
+                           pipeline::Runner::InvalidationMap &Invalidations);
+
+  virtual ~ModelInvalidationRule() = default;
+
+private:
+  static std::vector<ModelInvalidationRule *> &getRulesRegistry() {
+    static std::vector<ModelInvalidationRule *> Rules;
+    return Rules;
+  }
```

Comment:
Can you promote this to a `static` field?

Reply:
i could, but i prefer this way since this is resilient to initialization order fiasco

Reply:
Usually we solve it using `ManagedStatic`. I'd prefer to threat the same problem in the same way all over the codebase, it eases refactoring.